system/lib/compiler-rt/lib/asan/asan_mapping_emscripten.h

```diff
@@ -0,0 +1,86 @@
+//===-- asan_mapping_emscripten.h -------------------------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file is a part of AddressSanitizer, an address sanity checker.
+//
+// Emscripten-specific definitions for ASan memory mapping.
+//===----------------------------------------------------------------------===//
+#ifndef ASAN_MAPPING_EMSCRIPTEN_H
+#define ASAN_MAPPING_EMSCRIPTEN_H
+
+extern char __global_base;
+
+#define kLowMemBeg     ((uptr) &__global_base)
+#define kLowMemEnd     ((kLowShadowBeg << SHADOW_SCALE) - 1)
+
+#define kLowShadowBeg  0
+#define kLowShadowEnd  ((uptr) &__global_base - 1)
+
+#define kHighMemBeg    0
+
+#define kHighShadowBeg 0
+#define kHighShadowEnd 0
+
+#define kMidShadowBeg  0
+#define kMidShadowEnd  0
+
+#define kShadowGapBeg  (kLowMemEnd + 1)
+#define kShadowGapEnd  0xFFFFFFFF
+
+#define kShadowGap2Beg 0
+#define kShadowGap2End 0
+
+#define kShadowGap3Beg 0
+#define kShadowGap3End 0
+
+// The first 1/8 of the shadow memory space is wasted shadowing itself,
+// but that's better than introducing a branch to handle nullptr which will
+// wrap around and end up with no shadow.
```

Comment:
I wouldn't say the shadow memory shadowing the shadow memory is wasted, since the user could always do a stray write into the shadow memory that we would want to be able to catch. Actually, is the shadow shadow memory poisoned? If not, it might be worth poisoning.